### PR TITLE
Add Knowledge Graph functions to navigate to Documentation

### PR DESCRIPTION
**Pull request by MichaelDeng03**
Knowledge Graph: 
- Rename knowledge graph tool functions
- Add getters for knowledge graph nodes
- Separate node overall summary and child summaries to repo node sqlmodel model
- Add knowledge graph functions for node navigation

GH Daemon:
- Separate JWT Token and Installation Token from Daemon. 
- Add skeletons for calling Knowledge Graph to make commits.

### CODE REVIEW

File: concrete/db/orm/models.py [hunk 1]
```diff
@@ -263,13 +265,15 @@ class RepoNodeUpdate(NodeUpdate):
...
     summary: str | None = Field(description="Summary of the node.", default=None)
+    children_summaries: str | None = Field(description="Brief summary of each child node.", default=None)
     abs_path: str | None = Field(description="", default=None)
 
 
 class RepoNodeCreate(RepoNodeBase):
     pass
 
 
+# https://stackoverflow.com/questions/70958639/composite-indexes-sqlmodel
```
Comment (kgang) [inline concrete/db/orm/models.py:276]: Please include a descriptive phrase at minimum for links

File: concrete/operators.py [hunk 2]
```diff
@@ -215,15 +215,18 @@ def generate_summary(self, summary: str, implementation: str, *args, **kwargs) -
...
         """
         return f"""Given the following parent summary structure:
-Parent Summary: <{parent_summary}>
+Existing Parent Overall Summary: <{parent_summary}>
+Existing Parent Child Summaries: <{parent_child_summaries}>
 
-Your task is to update this summary with the new child summary:
+Your task is to update the existing summary with this new child summary:
 Child Name: <{child_name}>
 Child Summary: <{child_summary}>
```
Comment (kgang) [inline concrete/operators.py:231]: I wonder if structured data inside prompts should be JSON format cuz structure output models clearly know what they are on some level

File: webapp/daemons/server.py [hunk 3]
```diff
@@ -36,34 +37,102 @@ async def webhook_handler(self, request: Request):
...
+    @property
+    def token(self):
+        if not self._token or self._is_expired():
+            self._generate_jwt()
+        return self._token
+
+    def _is_expired(self):
+        return not self._expiry or self._expiry < time.time()
+
+    def _generate_jwt(self):
```
Comment (kgang) [inline webapp/daemons/server.py:69]: neat!

File: concrete/tools.py [hunk 4]
```diff
@@ -916,33 +962,167 @@ def _summarize_from_children(cls, repo_node_id: UUID) -> str:
...
+            with open(documentation_path, 'w') as f:
+                f.write('')
+
+        with open(documentation_path, 'r') as f:
+            existing_documentation = f.read()
+        with open(path, 'r') as f:
+            module_contents = f.read()
+
+        # Zero-shot w/ existing documentation + path content
+        from concrete.operators import Executive
```
Comment (kgang) [inline concrete/tools.py:1114]: why is this being imported here?

File: concrete/tools.py [hunk 5]
```diff
@@ -830,57 +850,79 @@ def _summarize_from_leaves(cls, root_node_id: UUID):
...
 
     @classmethod
-    def _summarize_leaf(cls, leaf_node_id: UUID) -> str:
+    def _upsert_leaf_summary(cls, leaf_node_id: UUID):
         """
-        Summarizes contents of a leaf node.
+        Creates or overwrites a leaf summary.
         TODO: Current implementation assumes that the leaf is a file - so generated summary is based on file contents.
-        Eventually, this should be able to summarize functions/classes in a file.
+        Eventually, this should be able to summ arize functions/classes in a file.
```
Comment (kgang) [inline concrete/tools.py:925]:
```suggestion
        Eventually, this should be able to summarize functions/classes in a file.
```

**Review by kgang — COMMENTED** [review]
looks good!

File: concrete/tools.py (hunk 4 above)
Comment (MichaelDeng03) [reply concrete/tools.py:1114]: I originally had imports at the function level for every function that required on Operator. Top-level imports caused a circular import. I'll move them to by tool/class level like pr93

File: concrete/operators.py (hunk 2 above)
Comment (MichaelDeng03) [reply concrete/operators.py:231]: That's a good thought, but probably not our immediate focus. I think we can examine prompt optimization more comprehensively when we look into dspy, context wrappers, and RAG

File: concrete/tools.py (hunk 4 above)
Comment (MichaelDeng03) [reply concrete/tools.py:1114]: Oh, actually I think i'll have to place them at the top of the function

**Review by kgang — APPROVED** [review]
LGTM!

File: concrete/db/orm/models.py [hunk 6]
```diff
@@ -248,6 +249,7 @@ class RepoNodeBase(Base):
     partition_type: str = Field(description="Type of the node. directory/file/chunk")
     name: str = Field(description="Name of the chunk. eg README.md, module.py/func_foo")
     summary: str = Field(description="Summary of the node.")
+    children_summaries: str = Field(description="Brief summary of each child node.")
```
Comment (abjjabjj) [inline concrete/db/orm/models.py:252]: is this a concatenation of summaries?

File: concrete/operators.py [hunk 7]
```diff
@@ -245,35 +249,38 @@ def update_parent_summary(self, parent_summary: str, child_summary: str, child_n
    ...
 
 Guidelines:
-- Ensure the Overall Summary provides a concise overview of all children.
+- Ensure the Overall Summary provides a high-level  overview of all children while preserving important implementation details.
```
Comment (abjjabjj) [inline concrete/operators.py:252]: nittiest of nits: is there supposed to be a double space between "high-level" and "overview"?

**Review by abjjabjj — COMMENTED** [review]
Looks good to me. Are the docs on GHDaemons up to date w.r.t. environment set up and execution?

File: concrete/operators.py (hunk 7 above)
Comment (kgang) [reply concrete/operators.py:252]: straight to jail

Comment (MichaelDeng03) [reply concrete/operators.py:252]: .... Did you have to scar my eyes after I merged? [](edited-with-Axolo)

Comment (kgang) [reply concrete/operators.py:252]: It's fine throwing up another PR [](edited-with-Axolo)

Comment (abjjabjj) [reply concrete/operators.py:252]: dang the timing is unfortunate [](edited-with-Axolo)

Comment (abjjabjj) [reply concrete/operators.py:252]: reviewing as you merged it lol [](edited-with-Axolo)

File: concrete/db/orm/models.py (hunk 6 above)
Comment (MichaelDeng03) [reply concrete/db/orm/models.py:252]: Yes its a concatenation of all child summaries. Its a little hacky, but I didn't want to bother too much with the implementation details.

**Pull request merged**